presto-main/src/main/java/com/facebook/presto/operator/project/MergingPageOutput.java

```diff
@@ -0,0 +1,175 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.project;
+
+import com.facebook.presto.spi.Page;
+import com.facebook.presto.spi.PageBuilder;
+import com.facebook.presto.spi.block.ArrayBlockBuilder;
+import com.facebook.presto.spi.type.Type;
+import com.google.common.collect.ImmutableList;
+import org.openjdk.jol.info.ClassLayout;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.NotThreadSafe;
+
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Queue;
+
+import static com.facebook.presto.spi.block.PageBuilderStatus.DEFAULT_MAX_PAGE_SIZE_IN_BYTES;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * This class is intended to be used right after the PageProcessor to ensure
+ * that the size of the pages returned by FilterAndProject and ScanFilterAndProject
+ * is big enough to do not introduce considerable synchronization overhead
+ * <p>
+ * As long as the input page contains more than {@link MergingPageOutput#DEFAULT_MIN_ROW_COUNT} rows
+ * or bigger than {@link MergingPageOutput#DEFAULT_MIN_PAGE_SIZE_IN_BYTES} it is returned as is without
+ * additional memory copy.
+ * <p>
+ * The page data that has been buffered so far before receiving a "big" page is being flushed
+ * before transferring a "big" page.
+ * <p>
+ * Although it is still possible that the {@link MergingPageOutput} may return a tiny page,
+ * this situation is considered to be rare due to the assumption that filter selectivity may not
+ * vary a lot based on the particular input page.
+ * <p>
+ * Considering the CPU time required to process(filter, project) a full (~1MB) page returned by a
+ * connector, the CPU cost of memory copying (< 50kb, < 1024 rows) is supposed to be negligible
+ */
+@NotThreadSafe
+public class MergingPageOutput
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(ArrayBlockBuilder.class).instanceSize();
+
+    private static final int DEFAULT_MIN_PAGE_SIZE_IN_BYTES = 50 * 1024;
+    private static final int DEFAULT_MIN_ROW_COUNT = 1024;
```

Comment:
I can try to decrease this constant as well as make it configurable. However for smaller constants we need to have #8663 merged all together with this patch. In case of high task concurrency ( e.g. 32 ) from 128 rows input page we will be getting 4 rows pages as a local partitioner output.